bgv/evaluator.go

```diff
@@ -10,455 +11,631 @@ import (
+//   - the level of opOut will be updated to min(op0.Level(), op1.Level())
+//   - the scale of opOut will be updated to op0.Scale * op1.Scale
+func (eval Evaluator) Mul(op0 *rlwe.Ciphertext, op1 rlwe.Operand, opOut *rlwe.Ciphertext) (err error) {
+
+	switch op1 := op1.(type) {
+	case rlwe.ElementInterface[ring.Poly]:
+
+		_, level, err := eval.InitOutputBinaryOp(op0.El(), op1.El(), 2, opOut.El())
+		if err != nil {
+			return fmt.Errorf("cannot Mul: %w", err)
+		}
+
+		opOut.Resize(opOut.Degree(), level)
 
-	ringT := eval.params.RingT()
+		if err = eval.tensorStandard(op0, op1.El(), false, opOut); err != nil {
+			return fmt.Errorf("cannot Mul: %w", err)
+		}
 
-	level := utils.MinInt(ctIn.Level(), ctOut.Level())
+	case *big.Int:
 
-	if ctIn.Scale.Cmp(eval.params.NewScale(1)) != 0 {
-		scalar = ring.BRed(scalar, ctIn.Scale.Uint64(), ringT.SubRings[0].Modulus, ringT.SubRings[0].BRedConstant)
-	}
+		_, level, err := eval.InitOutputUnaryOp(op0.El(), opOut.El())
+		if err != nil {
+			return fmt.Errorf("cannot Mul: %w", err)
+		}
 
-	scalarBig := new(big.Int).SetUint64(scalar)
+		opOut.Resize(op0.Degree(), level)
 
-	scalarBig.Mul(scalarBig, eval.tInvModQ[level])
+		ringQ := eval.parameters.RingQ().AtLevel(level)
 
-	eval.params.RingQ().AtLevel(level).AddScalarBigint(ctIn.Value[0], scalarBig, ctOut.Value[0])
+		TBig := eval.parameters.RingT().ModulusAtLevel[0]
 
-	if ctIn != ctOut {
-		for i := 1; i < ctIn.Degree()+1; i++ {
-			ring.Copy(ctIn.Value[i], ctOut.Value[i])
+		op1.Mod(op1, TBig)
+
+		// If op1 > T/2 then subtract T to minimize the noise
+		if op1.Cmp(new(big.Int).Rsh(TBig, 1)) == 1 {
+			op1.Sub(op1, TBig)
 		}
 
-		ctOut.MetaData = ctIn.MetaData
-	}
-}
+		for i := 0; i < op0.Degree()+1; i++ {
+			ringQ.MulScalarBigint(op0.Value[i], op1, opOut.Value[i])
+		}
 
-// AddScalarNew adds a scalar to ctIn and returns the result in a new ctOut.
-func (eval *evaluator) AddScalarNew(ctIn *rlwe.Ciphertext, scalar uint64) (ctOut *rlwe.Ciphertext) {
-	ctOut = NewCiphertext(eval.params, ctIn.Degree(), ctIn.Level())
-	eval.AddScalar(ctIn, scalar, ctOut)
-	return
-}
+	case uint64:
+		return eval.Mul(op0, new(big.Int).SetUint64(op1), opOut)
+	case int:
+		return eval.Mul(op0, new(big.Int).SetInt64(int64(op1)), opOut)
+	case int64:
+		return eval.Mul(op0, new(big.Int).SetInt64(op1), opOut)
+	case []uint64, []int64:
+
+		_, level, err := eval.InitOutputUnaryOp(op0.El(), opOut.El())
+		if err != nil {
+			return fmt.Errorf("cannot Mul: %w", err)
+		}
 
-// MulScalar multiplies ctIn with a scalar and returns the result in ctOut.
-func (eval *evaluator) MulScalar(ctIn *rlwe.Ciphertext, scalar uint64, ctOut *rlwe.Ciphertext) {
-	ringQ := eval.params.RingQ().AtLevel(utils.MinInt(ctIn.Level(), ctOut.Level()))
-	for i := 0; i < ctIn.Degree()+1; i++ {
-		ringQ.MulScalar(ctIn.Value[i], scalar, ctOut.Value[i])
+		// Instantiates new plaintext from buffer
+		pt, err := rlwe.NewPlaintextAtLevelFromPoly(level, eval.buffQ[0])
+		if err != nil {
+			panic(err)
```

Comment:
error expected here ? or never suppose to happen ?

Reply:
It is not supposed to happen, unless the user tempers with the internal buffers (which is possible since they are public). And in this case a panic is appropriate.

Reply:
Added comments on panics